conseil-api/src/main/scala/tech/cryptonomic/conseil/api/ConseilApi.scala

```diff
@@ -0,0 +1,136 @@
+  extends EnableCORSDirectives {
+
+  private val transformation = new UnitTransformation(config.metadata)
+  private val cacheOverrides = new AttributeValuesCacheConfiguration(config.metadata)
+
+  private val apiCache = new ApiCache()
+
+  private implicit val mat: ActorMaterializer = ActorMaterializer()
+  private implicit val dispatcher: ExecutionContext = system.dispatcher
+  private implicit val contextShift: ContextShift[IO] = IO.contextShift(dispatcher)
+
+  config.nautilusCloud.foreach { ncc =>
+    system.scheduler.schedule(ncc.delay, ncc.interval)(Security.updateKeys(ncc))
+  }
+
+  // this val is not lazy to force to fetch metadata and trigger logging at the start of the application
+  private val metadataService =
+    new MetadataService(config.platforms, transformation, cacheOverrides, apiCache.cachedDiscoveryOperations)
+
+  lazy val route: Route = {
+    lazy val platformDiscovery = PlatformDiscovery(metadataService)
+
+    lazy val routeDirectives = new RecordingDirectives
+    lazy val securityDirectives = new ValidatingDirectives(config.security)
+
+    import routeDirectives._
+    import securityDirectives._
+
+    concat(
+      pathPrefix("docs") {
+        pathEndOrSingleSlash {
+          getFromResource("web/index.html")
+        }
+      },
+      pathPrefix("swagger-ui") {
+        getFromResourceDirectory("web/swagger-ui/")
+      },
+      Docs.route,
+      cors() {
+        enableCORS {
+          implicit val correlationId: UUID = UUID.randomUUID()
+          handleExceptions(loggingExceptionHandler) {
+            extractClientIP {
+              ip =>
+                recordResponseValues(ip)(mat, correlationId) {
+                  timeoutHandler {
+                    concat(
+                      validateApiKey { _ =>
+                        concat(
+                          logRequest("Conseil", Logging.DebugLevel) {
+                            AppInfo.route
+                          },
+                          logRequest("Metadata Route", Logging.DebugLevel) {
+                            platformDiscovery.route
+                          },
+                          apiCache.cachedDataEndpoints(metadataService)
+                            .map {
+                              case (platform, routes) =>
+                                logRequest(s"$platform Data Route", Logging.DebugLevel) {
+                                  routes.getRoute ~ routes.postRoute
+                                }
+                            }
+                            .reduce(_ ~ _)
+                        )
+                      },
+                      options {
+                        // Support for CORS pre-flight checks.
+                        complete("Supported methods : GET and POST.")
+                      }
+                    )
+                  }
+                }
+            }
+          }
+        }
+      }
+    )
+  }
+
+  private class ApiCache {
```

Comment:
Wouldn't an `object` serve the same purpose more clearly?

Reply:
Here we need more docs to be able to easily make out what is what, especially with String map keys.
Can't existing wrappers be used for keys? Especially Value classes (if there is any...)

Reply:
So, here we are using `BlockchainPlatform` to fill up the map. Mostly because it's sealed trait, which provides all possible cases (and warning is displayed when any of it is not implemented in `forVisiblePlatforms` method, but... the `MetadataService` actually uses `Platform` case class, which does not have such option. 

That being said, I can use the same type here (which is `Platform`) but loose this warning and compile-time, or use `BlockchainPlatform` but when, `MetadataService` would need to convert `Platform` -> `BlockchainPlatfrom` every time.

Probably both works, question is which one is better option?

Reply:
I'm not sure I see what you mean here, was I unclear, possibly? I'll restate my observations:

1. why `private class ApiCache` vs. `private object ApiCache`?
2. in `cachedDiscoveryOperations` and `cachedDataEndpoints` the maps have both Strings as keys. Is there any more explicitly typed wrapper to use, to understand what should be expected there? The latter is because I needed to check both here and where those maps are used, to understand the logic, and it wasn't immediately obvious to me.

Maybe the original code was lacking in terms of comments, it's personally something I try to recommend always, to explain any piece of code which might not be obvious, YMMV.

Reply:
Point 1. is clear to me. Point 2 is a bit more complicated. I just explained above why `Map[String, ...]` not `Map[anything-else, ...]`. Mostly because it is hard to find the best option to pick, because we have two different types describing kind of the same (or even 3 very similar): `BlockchainPlatform`, `Platform` and `PlatformPath`.

I just said, that the best idea would be to change `Map[String, ...]` into `Map[BlockchainPlatform, ...]`, but then, since we are accessing these data from `MetadataService` and this service has `PlatformPath`, I would need to convert `PlatformPath` -> `BlockchainPlatform` every time someone would like to access these services. The common ground for `BlockchainPlatform` and `PlatformPath` is Platform's name. That's why `String` :)

I was just asking, taking into what I just wrote, if you see, which type from what we have, would work better, because for me, `String` is such type, at least for the time being.

Reply:
I trust your decision on this, I see that it's more complex than it appeared at first sight